server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java

```diff
@@ -165,6 +178,69 @@ public void messageReceived(UpdateVisibleCheckpointRequest request, TransportCha
         }
     }
 
+    private class GetMergedSegmentFilesRequestHandler implements TransportRequestHandler<GetSegmentFilesRequest> {
+        @Override
+        public void messageReceived(GetSegmentFilesRequest request, TransportChannel channel, Task task) throws Exception {
+            ActionListener<GetSegmentFilesResponse> listener = new ChannelActionListener<>(
+                channel,
+                Actions.GET_MERGED_SEGMENT_FILES,
+                request
+            );
+            final ShardId shardId = request.getCheckpoint().getShardId();
+            IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+            IndexShard indexShard = indexService.getShard(shardId.id());
+            if (indexShard.state().equals(IndexShardState.STARTED) == false || indexShard.isPrimaryMode() == false) {
+                throw new IllegalStateException(String.format(Locale.ROOT, "%s is not a started primary shard", shardId));
+            }
+
+            RemoteSegmentFileChunkWriter mergedSegmentFileChunkWriter = new RemoteSegmentFileChunkWriter(
+                request.getReplicationId(),
+                indexShard.getRecoverySettings(),
+                new RetryableTransportClient(
+                    transportService,
+                    request.getTargetNode(),
+                    indexShard.getRecoverySettings().getMergedSegmentReplicationTimeout(),
+                    logger
+                ),
+                request.getCheckpoint().getShardId(),
```

Comment:
some bit of duplication here and in get segments handler. see if this can be deduped as well.